### PR TITLE
Model Implementation

### PR DESCRIPTION
**Pull request by djr277**
I've added models for BusPath, Place/PlaceResult, BusLocation, Direction, and Waypoint. Not all of the models are complete yet, but they should provide enough of a basis for us to start working on the Controllers.

### CODE REVIEW

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusLocation.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
+package ithaca_transit.android.cornellappdev.com.ithaca_transit.Models
+
+import android.widget.ImageView
+import com.google.android.gms.maps.model.Marker
+import java.util.*
+
+enum class BusDataType {
+    noData, validData, invalidData
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusLocation.kt]: should be alphabetized

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusLocation.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
+package ithaca_transit.android.cornellappdev.com.ithaca_transit.Models
+
+import android.widget.ImageView
+import com.google.android.gms.maps.model.Marker
+import java.util.*
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusLocation.kt]: not good practice to ever import `*`, can you just import what's necessary?

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusLocation.kt [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+enum class BusDataType {
+    noData, validData, invalidData
+}
+
+class BusLocation(dataType: BusDataType, destination: String, deviation: Int, delay: Int,
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusLocation.kt]: `delay` should be before `destination`, looks like you were trying to alphabetize this

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusLocation.kt [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        this.routeID = routeID
+        this.runID = runID
+        this.speed = speed
+        this.tripID = tripID
+        this.vehicleID = vehicleID
+    }
+
+    fun routeNumer(): Int {
+        return if (routeID != null) Integer.parseInt(routeID) else 0
+
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusLocation.kt:60]: delete newline

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusPath.kt [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    init {
+        this.waypoints = waypoints
+        this.traveledPolyline = PolylineOptions()
+        this.color = Color.rgb(0,0,0)
+    }
+}
+
+class BusPath(waypoints: Array<Waypoint>) : Path(waypoints) {
+
+    var dashLengths : Array<Int> = arrayOf(6, 4)
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusPath.kt]: this should probably be defined in constants up top, it's not clear what 6 and 4 represent

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusPath.kt [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    private var dashColors : Array<Int> = emptyArray()
+    private var polylineWidth : Float?
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusPath.kt]: you have spaces before `:` in this line and the two above it

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusPath.kt [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    private var traveledPath: List<LatLng> ?= null
+    private var untraveledPath: List<LatLng> ?= null
+
+    init{
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusPath.kt]: space after `init`

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusPath.kt [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        createPath(waypoints)
+        this.untraveledPath = super.traveledPolyline!!.points
+        this.traveledPath = untraveledPath
+        super.traveledPolyline!!.color(super.color)
+        super.traveledPolyline!!.width(polylineWidth!!)
+    }
+
+    fun createPath(waypoints: Array<Waypoint>) {
+        var count = 0
+        for(point in waypoints)
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusPath.kt]: space after `for`, the `{` should be on the same line as the `for`:

I double checked [Kotlin reference](https://kotlinlang.org/docs/reference/control-flow.html) to make sure that this was idiomatic

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusPath.kt [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        {
+            this.traveledPolyline!!.add(point.coordinates())
+            count = count + 1
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusPath.kt:55]: `count += 1`?

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Direction.kt [hunk 10]
```diff
@@ -0,0 +1,68 @@
+package ithaca_transit.android.cornellappdev.com.ithaca_transit.Models
+
+import android.location.Location
+import com.google.android.gms.maps.model.LatLng
+import ithaca_transit.android.cornellappdev.com.ithaca_transit.R
+import java.util.*
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Direction.kt]: same comment regarding `*`

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Direction.kt [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+enum class DirectionType {
+    WALK, DEPART, ARRIVE, TRANSFER
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Direction.kt]: alphabetize

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Direction.kt [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+class Direction(type: DirectionType, name: String, startLocation: LocationObject, endLocation: LocationObject,
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Direction.kt]: alphabetize params

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Direction.kt [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+                startTime: Date, endTime: Date, routeNumer: Int, stops: Array<LocationObject>,
+                stayOnBusForTransfer: Boolean, tripIdentifiers: Array<String>, delay: Int, path: Array<LatLng>,
+                travelDistance: Double) {
+
+    private var type: DirectionType? = null 
+    private var name: String? = null 
+    private var startLocation: LocationObject? = null 
+    private var endLocation: LocationObject? = null 
+    private var startTime: Date? = null 
+    private var endTime: Date? = null
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Direction.kt]: alphabetize this section as well

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusPath.kt [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+import ithaca_transit.android.cornellappdev.com.ithaca_transit.R
+import java.security.AccessController.getContext
+
+enum class PathType {
+    driving, walking
+}
+
+open class Path(waypoints: Array<Waypoint>) {
+    var waypoints: Array<Waypoint> ?= null
+    var traveledPolyline: PolylineOptions ?= null
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusPath.kt]: alphabetize here

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusPath.kt [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    var color: Int
+
+    init {
+        this.waypoints = waypoints
+        this.traveledPolyline = PolylineOptions()
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusPath.kt:27]: alphabetize as well

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusLocation.kt [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+class BusLocation(dataType: BusDataType, destination: String, deviation: Int, delay: Int,
+                  direction: String, displayStatus: String, gpsStatus: Int, heading: Int,
+                  lastStop: String, lastUpdated: Date, latitude: Double, longitude: Double,
+                  name: Int, opStatus: String, routeID: String, runID: Int, speed: Int, tripID: String,
+                  vehicleID: Int) {
+
+    private var dataType: BusDataType? = null
+    private var destination: String? = null
+    private var deviation: Int? = null
+    private var delay: Int? = null
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusLocation.kt:19]: same comment regarding `delay`

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusPath.kt [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    var dashLengths : Array<Int> = arrayOf(6, 4)
+    private var dashColors : Array<Int> = emptyArray()
+    private var polylineWidth : Float?
+    private var traveledPath: List<LatLng> ?= null
+    private var untraveledPath: List<LatLng> ?= null
+
+    init{
+        super.color = ContextCompat.getColor(R.colors.blue);
+        this.dashColors = arrayOf(color!!, Color.rgb(0,0,0))
+        this.polylineWidth = 8.0F
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusPath.kt]: should be defined in a constant up top

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Direction.kt [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+        this.routeNumer = routeNumer 
+        this.stayOnBusForTransfer = stayOnBusForTransfer 
+        this.tripIdentifiers = tripIdentifiers 
+        this.delay = delay 
+        this.path = path 
+        this.stops = stops 
+        this.travelDistance = travelDistance 
+    }
+
+    //static variables
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Direction.kt]: space after `//`

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Direction.kt [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+
+    fun locationDescription(): String {
+        var instruct = ""
+        when (type) {
+            DirectionType.WALK -> instruct = "Walk to " + name
+            DirectionType.ARRIVE -> instruct = "Get off at " + name
+            DirectionType.DEPART -> instruct = "at " + name
+            DirectionType.TRANSFER -> instruct = "at " + name + ". Stay on bus"
+        }
+        return instruct
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Direction.kt]: why define a `instruct` variable that's `""` when you can `return` within each of the `->` cases?

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/LocationObject.kt [hunk 20]
```diff
@@ -0,0 +1,28 @@
+package ithaca_transit.android.cornellappdev.com.ithaca_transit.Models
+
+import com.google.android.gms.maps.model.LatLng
+
+class LocationObject(name: String, id: String, latitude: Double, longitude: Double){
+    var name: String ?= null
+    var id: String ?= null
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/LocationObject.kt:6]: same with alphabetizing

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/LocationObject.kt [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+    var latitude: Double ?= null
+    var longitude: Double ?= null
+
+    init{
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/LocationObject.kt]: space after `init`

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/LocationObject.kt [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+        this.name = name
+        this.id = id
+        this.latitude = latitude
+        this.longitude = longitude
+    }
+
+    constructor():this("","",0.0,0.0)
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/LocationObject.kt]:
```suggestion
    constructor(): this("", "", 0.0, 0.0)
```

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/LocationObject.kt [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+
+    constructor(name:String):this(name,"",0.0,0.0)
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/LocationObject.kt]:
```suggestion
    constructor(name: String): this(name, "", 0.0, 0.0)
```

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/LocationObject.kt [hunk 24]
```diff
@@ -0,0 +1,28 @@
...
+
+    constructor(name:String, latitude: Double, longitude: Double):this(name,"",latitude,
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/LocationObject.kt]:
```suggestion
    constructor(name: String, latitude: Double, longitude: Double): this(name, "", latitude,
```

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Place.kt [hunk 25]
```diff
@@ -0,0 +1,30 @@
+package ithaca_transit.android.cornellappdev.com.ithaca_transit.Models
+
+class Place {
+
+    var name: String? = null
+    var detail: String? = null
+    var placeID: String? = null
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Place.kt:7]: alphabetize

File: app/src/main/res/values/colors.xml [hunk 26]
```diff
@@ -1,5 +1,9 @@
 <?xml version="1.0" encoding="utf-8"?>
 <resources>
+    <color name="blue">#4a90e2</color>
+    <color name="red">#f2655d</color>
+    <color name="white">#ffffff</color>
+    <color name="green">#63c774</color>
```
Comment (young-k) [inline app/src/main/res/values/colors.xml]: this should be alphabetized, and you should either stick with `colorPrimary` naming conventions, or `blue`, but not mix and match both

File: app/src/main/res/values/strings.xml [hunk 27]
```diff
@@ -4,4 +4,6 @@
     <string name="title_dashboard">Dashboard</string>
     <string name="title_notifications">Notifications</string>
     <string name="title_activity_maps">Map</string>
+    <string name="location_walk">Walk to</string>
```
Comment (young-k) [inline app/src/main/res/values/strings.xml]: alphabetize

File: app/src/main/res/values/strings.xml [hunk 28]
```diff
@@ -4,4 +4,6 @@
...
+
```
Comment (young-k) [inline app/src/main/res/values/strings.xml]: delete

**Review by young-k — DISMISSED** [review]
Great work with this, are you using a linter? There seem to be a lot of minor style issues that should have been caught before PR.

In terms of commits, they aren't named according to our [conventions](https://bible.cornellappdev.com/standards/git) –– they should be in present tense, start with a capital letter, and unnecessary ones like `removed semicolons` should be squashed under other commits.

**Review by abdullah248 — APPROVED** [review]
Looking through it all looks good to me,  there may be some parts of models we can rework but I think it will become more clear as we move forward.

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Direction.kt (hunk 19 above)
Comment (young-k) [reply app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Direction.kt]: @abdullah248 can you address?

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/LocationObject.kt (hunk 22 above)
Comment (young-k) [reply app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/LocationObject.kt]: @abdullah248 unclear on why you resolved this as well?

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusLocation.kt [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+        this.name = name
+        this.opStatus = opStatus
+        this.routeID = routeID
+        this.runID = runID
+        this.speed = speed
+        this.tripID = tripID
+        this.vehicleID = vehicleID
+    }
+
+    fun routeNumer(): Int {
```
Comment (young-k) [inline app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusLocation.kt]: `routeNumber`

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Direction.kt (hunk 19 above)
Comment (abdullah248) [reply app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/Direction.kt]: Was updated to match suggestion

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/LocationObject.kt (hunk 22 above)
Comment (abdullah248) [reply app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/LocationObject.kt]: Was updated to match suggestion

File: app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusLocation.kt (hunk 29 above)
Comment (abdullah248) [reply app/src/main/java/ithaca_transit/android/cornellappdev/com/ithaca_transit/Models/BusLocation.kt]: Thanks

**Pull request merged**